web/packages/teleterm/src/ui/DocumentGatewayCliClient/DocumentGatewayCliClient.tsx

```diff
@@ -0,0 +1,158 @@
+  const [hasRenderedTerminal, setHasRenderedTerminal] = useState(false);
+
+  const gateway = clustersService.findGatewayByConnectionParams(
+    doc.targetUri,
+    doc.targetUser
+  );
+
+  // Once we render the terminal, we want to keep it visible. Otherwise removing the gateway would
+  // mean that this document would immediately unmount DocumentTerminal and close the PTY.
+  //
+  // After the gateway is closed, the CLI client will not be able to interact with the gateway
+  // target, but the user might still want to inspect the output.
+  if (gateway || hasRenderedTerminal) {
+    if (!hasRenderedTerminal) {
+      setHasRenderedTerminal(true);
+    }
+
+    return <DocumentTerminal doc={doc} visible={visible} />;
+  }
+
+  return <WaitingForGateway doc={doc} visible={visible} />;
+};
+
+const TIMEOUT_SECONDS = 10;
+
+const WaitingForGateway = (props: {
+  doc: types.DocumentGatewayCliClient;
+  visible: boolean;
+}) => {
+  const { doc, visible } = props;
+  const ctx = useAppContext();
+  const { documentsService } = useWorkspaceContext();
+  // If we depended on doc.status for hasTimedOut instead of using a separate state, then on reopen
+  // the doc would have status set to 'connected' on 'error' and it'd be updated from useEffect,
+  // meaning that there would be a brief flash of old state.
+  const [hasTimedOut, setHasTimedOut] = useState(false);
+
+  useEffect(() => {
+    // Update the doc state to make the progress bar show up in the tab bar.
+    // Once DocumentTerminal is mounted, it is going to update the status to 'connected' or 'error'.
+    documentsService.update(doc.uri, { status: 'connecting' });
+
+    const timeoutId = setTimeout(() => {
+      setHasTimedOut(true);
+      documentsService.update(doc.uri, { status: 'error' });
+    }, TIMEOUT_SECONDS * 1000);
+
+    return () => {
+      clearTimeout(timeoutId);
+    };
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
+
+  const openConnection = () => {
+    connectToDatabase(
+      ctx,
+      {
+        uri: doc.targetUri,
+        name: doc.targetName,
+        protocol: doc.targetProtocol,
+        dbUser: doc.targetUser,
+      },
+      { origin: 'reopened_session' }
+    );
+  };
+
+  return (
+    <Document visible={visible} px={2}>
+      <WaitingForGatewayContent
+        doc={doc}
+        hasTimedOut={hasTimedOut}
+        openConnection={openConnection}
+      />
+    </Document>
+  );
+};
+
+export const WaitingForGatewayContent = ({
+  doc,
+  hasTimedOut,
+  openConnection,
+}: {
+  doc: types.DocumentGatewayCliClient;
+  hasTimedOut: boolean;
+  openConnection: () => void;
+}) => (
+  <Flex gap={4} flexDirection="column" mx="auto" alignItems="center" mt={100}>
+    {hasTimedOut ? (
+      <div>
+        <StyledText>
+          A connection to <strong>{doc.targetName}</strong> as{' '}
+          <strong>{doc.targetUser}</strong> has not been opened up within{' '}
+          {TIMEOUT_SECONDS} seconds.
```

Comment:
I think it'd be better to have a more explicit error message here. Something like
```
Error connecting to {targetName} as {targetUser}: Connection timed out after {timeout_seconds} seconds.
```

I won't die on this hill tho. I remember talk about changing the way our error messages sound, but I don't remember if it was supposed to be more or less "machine-y".

Reply:
The message you proposed is more direct and uses the active voice which I like. However, in this particular case there's no connection that times out and there's no error. As I mentioned in the previous comment, all waiting here is fake – we assume that when the `DocumentGatewayCliClient` is open, there's `DocumentGateway` open as well. When the user restarts the app, we wait for `DocumentGateway` to create the gateway before `DocumentGatewayCliClient` attempts to connect to the gateway.

What's more, it might take `DocumentGateway` more than 10 seconds to open the gateway, in that case `DocumentGatewayCliClient` is going to switched from the "timed out" state to starting a terminal session automatically.

With that in mind, I don't know how I could transform the existing message from the passive voice to the active voice.